rpcserver: Fix websocket auth failure.

liukun opened the pull request:
Fix a bug that allow any user/password combinations except the correct one to pass the authentication of websocket and can act as a limited user.

~~I tried to add a test but the code is too deep in a long function to write a proper test. A more proper way is to reuse the function `func (s *Server) checkAuth()` in `rpcserver.go`, but that changed more lines and make the diff not clear enough. If need I can do the changes.~~

rstaudt2 reviewed (commented):
Good find!  This fix looks proper to me and is now the same as the logic for the [non-websocket case](https://github.com/decred/dcrd/blob/bcc5023fa94d2e0895682012081fbe3cef3d6b51/internal/rpcserver/rpcserver.go#L5884-L5890), which is correct.  It looks like it was unintentionally introduced in https://github.com/decred/dcrd/pull/2509 when changing the check to be constant time.

It also looks like this case can only be hit if an unauthenticated websocket connection is established to begin with since authenticated connections are [correctly rejected here](https://github.com/decred/dcrd/blob/bcc5023fa94d2e0895682012081fbe3cef3d6b51/internal/rpcserver/rpcserver.go#L6306-L6310) if an incorrect user/pass are provided.  This is likely why this wasn't noticed previously.

If you would like to add tests, you could consider doing it in a separate commit or PR from this fix if the concern was just that it is a larger diff.

On a minor note, per the [code contribution guidelines](https://github.com/decred/dcrd/blob/master/docs/code_contribution_guidelines.md#ModelGitCommitMessages), you should wrap your commit description to 72 chars and can also remove the PR checklist from the description now.

liukun commented:
@rstaudt2 Thanks! I refactored all the checks to use the function in `rpcserver.go` and added a test case. Not too big IMO, so I just commit this together. Also wrapped the commit message to 72 chars.

rstaudt2 reviewed (commented):
Creating a single function for this with corresponding tests makes sense to me—I left a few comments inline.

liukun commented on internal/rpcserver/rpcserver.go:
```diff
@@ -5878,21 +5878,34 @@ func (s *Server) checkAuth(r *http.Request, require bool) (bool, bool, error) {
+	return s.checkAuthHeader(authhdr[0], r.RemoteAddr)
+}
+
+// checkAuthHeader checks the HTTP Basic authentication string.
```
Added.

liukun commented on internal/rpcserver/rpcserver.go:
```diff
@@ -5878,21 +5878,34 @@ func (s *Server) checkAuth(r *http.Request, require bool) (bool, bool, error) {
 		// Request's auth doesn't match either user
-		log.Warnf("RPC authentication failure from %s", r.RemoteAddr)
+		log.Warnf("RPC authentication failure from %s", remoteAddr)
 		return false, false, errors.New("auth failure")
```
Removed the error.

liukun commented on internal/rpcserver/rpcserver_test.go:
```diff
@@ -170,3 +170,38 @@ func TestRpcServer(t *testing.T) {
+	}
+	if authed, _, err := s.checkAuthUserPass("", "", "wrong"); err == nil || authed {
+		t.Fatalf("should not pass auth")
+	}
```
Done.

rstaudt2 reviewed (commented):
The updates look good to me.  I noted just a couple of other small things inline.

liukun commented on internal/rpcserver/rpcserver.go:
```diff
@@ -5851,46 +5851,70 @@ func (s *Server) authMAC(dst, auth []byte) []byte {
+	authed = true
+	isAdmin = cmp == 1
+	return
+}
```
OK, I doubted the same thing when using named returns. Changed in the newest commit.

liukun commented on internal/rpcserver/rpcserver_test.go:
```diff
@@ -170,3 +170,69 @@ func TestRpcServer(t *testing.T) {
+	if err != nil {
+		t.Fatalf("unable to create RPC server: %v", err)
+	}
+	tests := []struct {
```
OK, renamed.

davecgh commented on internal/rpcserver/rpcserver.go:
```diff
@@ -5849,48 +5849,69 @@ func (s *Server) authMAC(dst, auth []byte) []byte {
 			log.Warnf("RPC authentication failure from %s",
 				r.RemoteAddr)
-			return false, false, errors.New("auth failure")
+			return false, false
```
This change is incorrect.  It means the calling code can no longer tell the difference between whether the authentication is required or not.  The Websockets endpoint must allow Websocket clients to authenticate either via HTTP basic authentication **or** via the [authenticate](https://github.com/decred/dcrd/blob/master/docs/json_rpc_api.mediawiki#authenticate) Websocket-only RPC as opposed to the header.

liukun replied:
The only effective change in this commit is https://github.com/liukun/dcrd/commit/e37ced291766227a32e037bb7b517ddbb5d16f6f . At first I only committed this one line change, then under the suggestions, I added helper functions and a test. No logic changes except the comparison `cmp|limitcmp != 0` is done.

liukun replied:
Removed this change in the newest commit. Add a test to check this behaviour.

davecgh commented on internal/rpcserver/rpcserver.go:
```diff
@@ -6291,8 +6312,8 @@ func (s *Server) route(ctx context.Context) *http.Server {
 		defer s.decrementClients()
-		_, isAdmin, err := s.checkAuth(r, true)
-		if err != nil {
+		authed, isAdmin := s.checkAuth(r, true)
```
Notice how this non-Websocket endpoint requires authentication via HTTP basic auth.

liukun replied:
Reverted.

davecgh commented on internal/rpcserver/rpcserver.go:
```diff
@@ -6303,8 +6324,8 @@ func (s *Server) route(ctx context.Context) *http.Server {
 	rpcServeMux.HandleFunc("/ws", func(w http.ResponseWriter, r *http.Request) {
-		authenticated, isAdmin, err := s.checkAuth(r, false)
-		if err != nil {
+		authenticated, isAdmin := s.checkAuth(r, false)
```
However, the Websocket endpoint does not.  These changes break that.

liukun replied:
OK, I will fix this.

liukun replied:
Reverted.

davecgh commented on internal/rpcserver/rpcserver.go:
```diff
@@ -5878,19 +5911,8 @@ func (s *Server) checkAuth(r *http.Request, require bool) (bool, bool, error) {
-	isAdmin := cmp == 1
-	return true, isAdmin, nil
+	authed, isAdmin := s.checkAuthHeader(authhdr[0], r.RemoteAddr)
+	return authed, isAdmin, nil
```
This is still incorrect.  This is not returning the error when the authentication fails, so the call to it in the non-websocket handler is now incorrect:

```Go
		_, isAdmin, err := s.checkAuth(r, true)
		if err != nil { // This will be `nil` even if the supplied credentials are incorrect....
			jsonAuthFail(w)
			return
		}
```

liukun replied:
Ah, fully understand now. Sorry. Added more lines in the test.

davecgh commented on internal/rpcserver/rpcserver.go:
```diff
@@ -5849,6 +5849,39 @@ func (s *Server) authMAC(dst, auth []byte) []byte {
+// The first bool return value signifies auth success (true if successful) and
+// the second bool return value specifies whether the user can change the state
+// of the server (true) or whether the user is limited (false).
+func (s *Server) checkAuthHeader(auth, remoteAddr string) (bool, bool) {
```
This doesn't really do what it's named since it's used for both the `authenticate` RPC and HTTP basic authentication paths.  I would say it should be named `checkAuthMAC`.

liukun replied:
Renamed.

davecgh commented on internal/rpcserver/rpcwebsocket.go:
```diff
@@ -1466,18 +1463,10 @@ out:
-				limitcmp := subtle.ConstantTimeCompare(mac, c.rpcServer.limitauthsha[:])
-				if cmp|limitcmp != 0 {
-					log.Warnf("Auth failure.")
+				c.authenticated, c.isAdmin = c.rpcServer.checkAuthUserPass(authCmd.Username, authCmd.Passphrase, c.addr)
```
This is exceeding the 80-char limit imposed by the code contribution guidelines.

davecgh commented on internal/rpcserver/rpcwebsocket.go:
```diff
@@ -1682,19 +1671,11 @@ out:
-							limitcmp := subtle.ConstantTimeCompare(authSha[:], c.rpcServer.limitauthsha[:])
-							if cmp != 1 && limitcmp != 1 {
-								log.Warnf("Auth failure.")
+							c.authenticated, c.isAdmin = c.rpcServer.checkAuthUserPass(authCmd.Username, authCmd.Passphrase, c.addr)
```
Same here regarding being too wide.

davecgh reviewed (changes requested):
The logic all looks correct now.  Thanks for the updates.  I have a few minor nits left inline and then it should be good for merge.

davecgh reviewed (approved):
Updates look good and I've tested the following scenarios:

* Connecting via HTTP without an auth header results in an authentication failure and immediate disconnection as expected
* Connecting via HTTP with an auth header that has incorrect credentials results in an authentication failure and immediate disconnection as expected
* Connecting via HTTP with an auth header that has correct credentials connects as expected
* Connecting via WebSockets without an auth header is allowed:
  * The `authenticate` RPC with invalid credentials results in an authentication failure and immediate disconnection as expected
  * The `authenticate` RPC with valid credentials results in a usable connection as expected
  * Any other RPCs aside from `authenticate` when not yet authenticated result in immediate disconnection as expected
* Connecting via WebSockets with an auth header that has incorrect credentials results in an authentication failure and immediate disconnection as expected
* Connecting via WebSockets with an auth header that has correct credentials connects as expected
